### ISSUE TITLE
Support Kotlin unsigned integer types

### ISSUE THREAD

**Issue by ForteScarlet** [babyfish-ct/jimmer#764]
你好！现在在 entity 中使用 Kotlin 的两个无符号整型（`UInt`, `ULong`）会出现异常，无法正常使用。

**Entity** :

```kotlin
@Entity
interface Building {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    val id: Long // `id bigserial`

    @Key
    val name: String

    val value: UInt
}
```

**Unit Test** 

```kotlin
@SpringBootTest(webEnvironment = MOCK)
@ActiveProfiles("...")
@AutoConfigureTestDatabase(replace = NONE)
class BuildingTests {
    @Autowired
    lateinit var sqlClient: KSqlClient

    @Test
    @Transactional
    @Rollback
    fun testBuild() {
        val new = Building {
            name = "New Building"
            value = 50u
        }

        val saved = sqlClient.save(new)
            .modifiedEntity

        println(saved)
    }
}
```

**Log**

```
Internal bug: Cannot find the getter of prop "value" of the interface "org.faraday.rs.entity.Building"
java.lang.AssertionError: Internal bug: Cannot find the getter of prop "value" of the interface "org.faraday.rs.entity.Building"
	at org.babyfish.jimmer.meta.impl.ImmutablePropImpl.<init>(ImmutablePropImpl.java:199)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$PropBuilder.build(ImmutableTypeImpl.java:1235)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$BuilderImpl.build(ImmutableTypeImpl.java:1160)
	at org.babyfish.jimmer.meta.impl.ImmutableTypeImpl$BuilderImpl.build(ImmutableTypeImpl.java:621)
	at org.faraday.rs.entity.BuildingDraft$$.<clinit>(BuildingDraft.kt:70)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/java.lang.reflect.Field.acquireFieldAccessor(Field.java:1185)
	at java.base/java.lang.reflect.Field.getFieldAccessor(Field.java:1164)
	at java.base/java.lang.reflect.Field.get(Field.java:442)
	at org.babyfish.jimmer.meta.impl.Metadata.create(Metadata.java:90)
	at org.babyfish.jimmer.impl.util.StaticCache.getWithoutLock(StaticCache.java:68)
	at org.babyfish.jimmer.impl.util.StaticCache.get(StaticCache.java:54)
	at org.babyfish.jimmer.meta.impl.Metadata.get(Metadata.java:28)
	at org.babyfish.jimmer.meta.ImmutableType.get(ImmutableType.java:26)
	at org.babyfish.jimmer.sql.runtime.EntityManager.<init>(EntityManager.java:62)
	at org.babyfish.jimmer.sql.runtime.EntityManager.fromResources(EntityManager.java:232)
	at org.babyfish.jimmer.sql.JSqlClientImpl$BuilderImpl.entityManager(JSqlClientImpl.java:1782)
	at org.babyfish.jimmer.sql.JSqlClientImpl$BuilderImpl.build(JSqlClientImpl.java:1573)
	at org.babyfish.jimmer.sql.di.JLazyInitializationSqlClient.sqlClient(JLazyInitializationSqlClient.java:35)
	at org.babyfish.jimmer.sql.di.AbstractJSqlClientDelegate.getEntities(AbstractJSqlClientDelegate.java:90)
	at org.babyfish.jimmer.sql.kt.impl.KSqlClientImpl.<init>(KSqlClientImpl.kt:76)
	at org.babyfish.jimmer.sql.kt.KSqlClientKt.toKSqlClient(KSqlClient.kt:1068)
	at org.babyfish.jimmer.spring.SqlClients.kotlin(SqlClients.kt:74)
	at org.babyfish.jimmer.spring.SqlClients.kotlin(SqlClients.kt:43)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig.kotlinSqlClient(SqlClientConfig.java:31)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$0.CGLIB$kotlinSqlClient$1(<generated>)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$FastClass$$1.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:348)
	at org.babyfish.jimmer.spring.cfg.SqlClientConfig$$SpringCGLIB$$0.kotlinSqlClient(<generated>)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:644)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:636)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:975)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:971)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
	at org.springframework.boot.test.context.SpringBootContextLoader.lambda$loadContext$3(SpringBootContextLoader.java:137)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:58)
	at org.springframework.util.function.ThrowingSupplier.get(ThrowingSupplier.java:46)
	at org.springframework.boot.SpringApplication.withHook(SpringApplication.java:1463)
	at org.springframework.boot.test.context.SpringBootContextLoader$ContextLoaderHook.run(SpringBootContextLoader.java:553)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:137)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:108)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:225)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:152)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:130)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:191)
	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:130)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:260)
	at org.springframework.test.context.junit.jupiter.SpringExtension.postProcessTestInstance(SpringExtension.java:163)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:735)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

简单调试发现: https://github.com/babyfish-ct/jimmer/blob/f8f768192338144cf6920f25cd2ded7f47011773/project/jimmer-core/src/main/java/org/babyfish/jimmer/meta/impl/ImmutablePropImpl.java#L194-L202

此处由于 `UInt` 是 `value class`, 在Java中会被编译为 `getValue-pVg5ArA` 而不是 `getValue`, 因此无法被寻得。

_也许可以考虑当 `kotlinClass` 或 `kotlinProp` 存在时通过Kotlin而不是Java获取 getter?  _

并且一些运算操作符似乎也不支持，例如

```kotlin
sqlClient.createQuery(Building::class) {
            where(table.id.plus(3) eq 1) // 普通数字可以 plus
            // where(table.value.plus(3u) eq 1u) // 不支持 plus 操作符

           // 也许可以通过native sql临时处理，只不过因为上述错误暂时无法测试
            where(
                sql(UInt::class, "%e + %v") {
                    expression(table.value)
                    value(3u)
                }.eq(1u)
            )
            select(table)
        }.execute()
```

**Comment by Enaium** [comment]
无符号类型还不是正式版本

**Comment by ForteScarlet** [comment]
> 无符号类型还不是正式版本

我想表述的可能是指 [无符号整型](https://kotlinlang.org/docs/unsigned-integer-types.html) ，就是 `UInt` `ULong` 这些。
不考虑数组，普通的无符号整型应该已经在 Kotlin 1.5 就 [稳定](https://kotlinlang.org/docs/whatsnew15.html#stable-unsigned-integer-types) 了。

> 以及由此错误延伸，可能所有的 value classes 都无法被支持。也许需要一个独立的 issue 记录一下 ?

**Comment by babyfish-ct** [comment]
Try `0.9.12`

The kotlin value types *(not only Uxxx)* is complex, now jimmer forbid properties value type by compilation error to protect itself before unified logic is implementation if the feature

**Comment by ForteScarlet** [comment]
Thx! A clear error does prevent this problem in advance as well.
But does this mean that the current issue is not actually resolved? Do we need to close this issue or keep it op until this issue is actually resolved in the future? 🤔

This does not mean that the issue needs to be resolved urgently, but merely serves as a memo and a way for others to learn about the issue.

**Comment by babyfish-ct** [comment]
Yes, this issue indicates that kotlin value types *(not only Uxxx types)* can broken the destroys the architecture. 

This is a small feature, but it would be costly to uniformly solve this problem and should not be considered by the 0.x era framework.

**Comment by babyfish-ct** [comment]
The JVM doesn't support custom value types, and before that, Kotlin couldn't help itself and introduced compile-time magic, which brought terrible unnecessary behind-the-scenes complexity. Frankly, I don't know if it is worth supporting even in the long run.

**Comment by ForteScarlet** [comment]
Yes. `value class` is very complex especially to juggle Java and Kotlin.
I mean, do we need to re-open this issue or document this with a new issue? Though it doesn't need to be fixed right away or delayed until after `1.0` .

It's just used to tell people who come after them: this is a planned feature that doesn't need to be mentioned again.

**Comment by ForteScarlet** [comment]
Oh, and if there's no clear plan, then it's fine not to reopen it. It all depends on your plan.
